clippy_lints/src/xor_used_as_pow.rs

```diff
@@ -0,0 +1,54 @@
+use crate::utils::span_lint_and_sugg;
+use if_chain::if_chain;
+use rustc::lint::{EarlyContext, EarlyLintPass, LintArray, LintPass};
+use rustc::{declare_lint_pass, declare_tool_lint};
+use rustc_errors::Applicability;
+use syntax::ast::{BinOpKind, Expr, ExprKind, LitKind};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for use of `^` operator when exponentiation was intended.
+    ///
+    /// **Why is this bad?** This is most probably a typo.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust,ignore
+    /// // Bad
+    /// 2 ^ 16;
+    ///
+    /// // Good
+    /// 1 << 16;
+    /// 2i32.pow(16);
+    /// ```
+    pub XOR_USED_AS_POW,
+    correctness,
+    "use of `^` operator when exponentiation was intended"
+}
+
+declare_lint_pass!(XorUsedAsPow => [XOR_USED_AS_POW]);
+
+impl EarlyLintPass for XorUsedAsPow {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &Expr) {
+        if_chain! {
+            if let ExprKind::Binary(op, left, right) = &expr.node;
+            if BinOpKind::BitXor == op.node;
+            if let ExprKind::Lit(lit) = &left.node;
+            if let LitKind::Int(2, _) = lit.node;
+            if let ExprKind::Lit(lit) = &right.node;
+            if let LitKind::Int(right, _) = lit.node;
+            then {
+                span_lint_and_sugg(
+                    cx,
+                    XOR_USED_AS_POW,
+                    expr.span,
+                    "`^` is not an exponentiation operator but was used as one",
+                    "did you mean to write",
+                    format!("1 << {}", right),
```

Comment:
Should the suggestion be about `.pow()` instead? The premise is, quote, "`^` is not an exponentiation operator but was used as one", so it might make more sense to suggest an actual exponentiation operator than a seemingly completely unrelated bit shift?

Reply:
`2^x == 1 << x` (That's only true for powers of 2). You don't want to use `pow` for powers of 2, because of performance reasons.

If we add a suggestion for `10^x`, the lint should obviously suggest `10.pow(x)`.

Reply:
Nod, I'm aware of how the bitshift works for powers of two, was just wondering if that makes for a somewhat confusing suggestion - telling the user they probably intended exponentiation, and then suggesting a bitshift. Especially, but not limited to, if the mistake is made by a beginner, the suggestion might seem strange 🤔

Reply:
Good point! I would suggest the bitshift, but write this in the documentation of the lint.

Reply:
(Additionally, arguably, why should bitshift be more efficient than `.pow()`, especially for known-at-compile-time values, shouldn't that be a compiler optimization that the programmer doesn't need to worry about? 🤔 )

Reply:
Another interesting thought here for consideration, perhaps recommending the `std::_::MAX` values, if applicable 🤔  https://github.com/rust-lang/rust/issues/61934#issue-457531302

Reply:
> shouldn't that be a compiler optimization

Yeah it should be one for `pow(some_power_of_two, x)`. But using a bitshift is always performant (and a pretty standard thing to do). Relying on the compiler might not always produce performant code. And Clippy should always suggest the best possible code IMO.

Reply:
We don't need an extra lint for this, I think. But you can do something like:
```rust
// NOTE: This is pseudo code and the messages need improvement. ;)
if lhs == 2 {
    span_lint(LINT, "this looks like you want to pow", "try bitshift");
} else {
    span_lint(LINT, "..", "try pow");
}
```

Reply:
Yeah, that's akin to what I was thinking. I think I'll also look at incorporating Walther's suggestion (and the comment from [this thread](https://github.com/rust-lang/rust/issues/61934#issue-457531302)) to recommend `std::_::MAX` in the case of rhs being an Int literal of 8, 16, 32, or 64.